### PR TITLE
Implement `Digest` class in `Digest::CRC32` and `Digest::Adler32`

### LINKED ISSUE #10343: Digest::Adler32 and Digest::CRC32 do not inherit from ::Digest
The `::Digest` abstract class represents an API for calculating/updating/finalizing digests (originally based off of Ruby's `Digest::Base` class). However, `Digest::Adler32` and `Digest::CRC32` do not inherit from `::Digest`. However, `OpenSSL::Digest`, `Crystal::Digest::MD5`, `Crystal::Digest::SHA1` do inherit from `::Digest`.

I feel like every type of digest algorithm under `Digest::` should inherit from `::Digest` and implement it's API.

### PR DESCRIPTION
**Pull request by BlobCodes**
Implements the Digest methods in Digest::CRC32 and Digest::Adler32.
Fixes #10343

### CODE REVIEW

**Comment by potomak** [comment]
LGTM

I’m wondering why the PR doesn’t have a review. Maybe try to ping a member of the core team?

File: src/digest/adler32.cr [hunk 1]
```diff
@@ -19,4 +27,27 @@ module Digest::Adler32
...
+  def update_impl(data : Bytes) : Nil
+    @digest = Adler32.update(data, @digest)
+  end
+
+  # :nodoc:
+  def final_impl(dst : Bytes) : Nil
+    dst[0] = (@digest >> 24).to_u8!
+    dst[1] = (@digest >> 16).to_u8!
+    dst[2] = (@digest >> 8).to_u8!
+    dst[3] = (@digest).to_u8!
```
Comment (straight-shoota) [inline src/digest/adler32.cr:41]: There is a dedicated method for big endian encoding which we can use here.

```suggestion
    IO::ByteFormat::BigEndian.encode(@digest, dst)
```

It's probably also little bit more efficient, using a single `memcpy`.

File: src/digest/crc32.cr [hunk 2]
```diff
@@ -19,4 +27,27 @@ module Digest::CRC32
...
+  def update_impl(data : Bytes) : Nil
+    @digest = CRC32.update(data, @digest)
+  end
+
+  # :nodoc:
+  def final_impl(dst : Bytes) : Nil
+    dst[0] = (@digest >> 24).to_u8!
+    dst[1] = (@digest >> 16).to_u8!
+    dst[2] = (@digest >> 8).to_u8!
+    dst[3] = (@digest).to_u8!
```
Comment (straight-shoota) [inline src/digest/crc32.cr:41]: ditto:
```suggestion
    IO::ByteFormat::BigEndian.encode(@digest, dst)
```

**Review by straight-shoota — APPROVED** [review]

File: src/digest/adler32.cr (hunk 1 above)
Comment (BlobCodes) [reply src/digest/adler32.cr:41]:
```
[blobcodes@toolbox mangaba]$ crystal eval '
require "benchmark"

bytes_store = uninitialized UInt8[4]
bytes = bytes_store.to_slice
n = rand(UInt32::MAX)

Benchmark.ips do |x|
  x.report("after") { IO::ByteFormat::BigEndian.encode(n, bytes) }
  x.report("before") do
    bytes[0] = (n >> 24).to_u8!
    bytes[1] = (n >> 16).to_u8!
    bytes[2] = (n >> 8).to_u8!
    bytes[3] = (n).to_u8!
  end
end

# Avoid LLVM optimizations by actually using the result
puts bytes
' --release
 after 173.22M (  5.77ns) (± 0.64%)  0.0B/op   1.55× slower
before 269.07M (  3.72ns) (± 0.84%)  0.0B/op        fastest
Bytes[62, 20, 67, 3]
```

It is slower and the same pattern is also used in other places:

https://github.com/crystal-lang/crystal/blob/d72f6eaf7b42ab32d0ce3f9e64d09f0c0509e089/src/crypto/bcrypt.cr#L141-L144

**Review by HertzDevil — APPROVED** [review]

**Comment by straight-shoota** [comment]
This came in too late for 1.6.0. I don't think we can merge it now far within the freeze period. It might cause some trouble, so it's better for 1.7.0.

**Pull request merged**